src/databricks/labs/ucx/hive_metastore/catalog_schema.py

```diff
@@ -50,32 +51,20 @@ def _get_catalog_schema_grants(self):
         # filter on grants to only get database level grants
         database_grants = [grant for grant in grants if grant.table is None and grant.view is None]
         for db_grant in database_grants:
-            new_grants.append(
-                dataclasses.replace(
-                    db_grant,
-                    # replace source database with taget UC database
-                    database=src_trg_schema_mapping[db_grant.database]['target_schema'],
-                    # replace hive_metastore with target UC catalog
-                    catalog=src_trg_schema_mapping[db_grant.database]['target_catalog'],
-                )
-            )
+            for target_catalog, target_schema in src_trg_schema_mapping[db_grant.database]:
+                new_grants.append(dataclasses.replace(db_grant, catalog=target_catalog, database=target_schema))
         for grant in new_grants:
             catalog_grants.add(dataclasses.replace(grant, database=None))
         new_grants.extend(catalog_grants)
         return new_grants
 
-    def _get_database_source_target_mapping(self) -> dict[str, dict]:
-        """generate a dictionary of source database in hive_metastore and its
-        mapping of target UC catalog and schema from the table mappings."""
-        src_trg_schema_mapping: dict[str, dict] = {}
+    def _get_database_source_target_mapping(self) -> dict[str, set[tuple[str, str]]]:
```

Comment:
since the tuple is catalog & schema, can we use `catalog.SchemaInfo` from sdk instead of `tuple[str, str]`

Reply:
good, this makes it better! Rewrote it to a dict with list values, as the SchemaInfo is non hashable, which is required for the set